### DISCUSSION TITLE
Custom cookie creation (size limitation workaround)

### DISCUSSION THREAD

**Discussion by digitalhank** [nextauthjs/next-auth#2628] [Ideas]
## Summary 💭

Due to the size limitations of cookies, i cannot store both the refresh & access token i am receiving from Cognito in the session cookie. It would be incredibly favourable if the library allowed you to a create cookies arbitrarily so that i for instance, could store the refresh token inside a separate cookie.

Edit - i realize this is fairly trivial to implement yourself but it would be nice if the library offered this feature oob.

**Comment by balazsorban44** [comment]
Thank you for this issue. To overcome this, currently you could utilize a database adapter. We offer quite a few options:

https://next-auth.js.org/adapters/overview

Would be also interesting to know if you are interested in creating a PR for this with your ideal implementation? 🙂

Ideally you could fill out the entire issue template you probably skipped, so others can evaluate your suggestion as well.

**Comment by mtt87** [comment]
Out of curiosity how long are the cognito `access_token` and `refresh_token`?
I'm using another oAuth provider and I can easily fit both in the session cookie to be honest.

**Comment by digitalhank** [comment]
@balazsorban44 thank you but unfortunately our service does not store any data.

**Comment by digitalhank** [comment]
@mtt87 refresh token is about 1700 b and the session token, which consists of a few profile properties & the access token is around 1850. total size is below chrome's limit but the browser was still printing warnings.

**Comment by balazsorban44** [comment]
would be interesting to know what is inside the access token? We use IdentityServer 4, and by not adding identity claims to the access token, we could spare a lot of space.

Could you share your `[...nextauth].js` file?

**Comment by digitalhank** [comment]
see here - https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-the-access-token.html

**Comment by dennervidal** [comment]
@digitalhank can you share your current solution for this? I'm facing the same issue using keycloak provider

**Reply by TonyBrobston** [reply]
> keycloak

I'm also using Keycloak and having the same issue. I have no control over how large the tokens are. I unfortunately need to hang on to the `access_token` and the `refresh_token`.

I plan to try the custom cookie approach described here in the next comment (https://github.com/nextauthjs/next-auth/discussions/2628#discussioncomment-1256545).

**Reply by TonyBrobston** [reply]
I gave this a go. What I've since realized is what I had originally thought; storing tokens in `next-auth` session or token or storing any of these tokens in a cookie makes them available to the browser (I'm sure this is obvious to others, but I wanted to check before I assumed anything). If a user gets ahold of an `access_token` they can call our api's in unwanted ways, so we prefer to keep this out of the browser. There's a similar situation with `refresh_token`, they could use the `refresh_token` to get an `access_token` and do the same. We will likely implement a cache on the backend.

**Reply by digitalhank** [reply]
You should not allow sensitive cookies to be accessed by the client. Utilise the `httpOnly: true` flag when creating your custom cookie. The cookies generated by the library already do this.

**Reply by digitalhank** [reply]
Additionally, i would avoid retrieving sensitive data when invoking getSession(). This is unavoidable if your app is purely client based & the services you depend on reside outside your domain. If that is not the case, I would recommend either proxying your calls through a nextjs api route or hosting a middlelayer service (think bff) on the same domain as your client.

**Reply by balazsorban44** [reply]
Perfect explanation by @digitalhank.

**Reply by TonyBrobston** [reply]
> You should not allow sensitive cookies to be accessed by the client. Utilise the `httpOnly: true` flag when creating your custom cookie. The cookies generated by the library already do this.

I'm not sure I follow. How do I not allow cookies to be accessed by the client?

Here is code for setting the cookie:
```
cookies.set('refresh_token', account.refresh_token, {
    expires: date,
    httpOnly: true,
    overwrite: true,
    sameSite: 'strict',
});
```
I have `httpOnly: true` set, however I see the `refresh_token` in Chrome Developer tools -> Application -> Cookies -> my local url

Am I doing something wrong?

**Reply by digitalhank** [reply]
Your configuration looks fine. The Chrome dev tools allow you to inspect each cookie attached to your instance, that doesn't mean your client is able to access said cookies. Try running `await window.cookieStore.getAll()` in your Chrome dev console and you'll see exactly what cookies your client can access.

A server on the same domain would be able to access every cookie you see in Application -> Cookies.

Read more about cookies [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies)

**Comment by digitalhank** [comment]
@dennervidal something like this 
```
import NextAuth from 'next-auth'
import Cookies from 'cookies'

const setRefreshCookie = ({ cookies, refresh_token }) => {
    const date = new Date()
    const time = date.getTime()
    const expireTime = time + 24 * 60 * 60 * 1000 * 30 // 30 days
    date.setTime(expireTime)

    cookies.set('refresh_token', refresh_token,
        {
            sameSite: 'strict',
            overwrite: true,
            expires: date,
            httpOnly: true
        })
}

const Auth = (req, res) => {
    const cookies = new Cookies(req, res)

    return (
        NextAuth(req, res, {
            secret: process.env.JWT_SECRET,
            callbacks: {
                async session(session, token) {
                    if (token) {
                        session.user.last_name = token.last_name
                        session.user.first_name = token.first_name
                        session.user.role = token.role
                        session.access_token = token.access_token
                        session.error = token.error
                    }

                    return session
                },
                async jwt(token, user, account, profile) {
                    if (profile) {
                        token.last_name = profile['custom:last_name']
                        token.first_name = profile['custom:first_name']
                        token.role = profile['custom:role']
                    }

                    if (account) {
                        token.access_token = account.access_token
                        token.accessTokenExpires = Date.now() + account.expires_in * 1000

                        setRefreshCookie({ cookies, refresh_token: account.refresh_token })
                    }

                    if (Date.now() < token.accessTokenExpires) {
                        return token
                    }
                }
            },
            providers: [
                Providers.Cognito({
                    clientId: process.env.COGNITO_CLIENT_ID,
                    clientSecret: process.env.COGNITO_CLIENT_SECRET,
                    domain: publicRuntimeConfig.COGNITO_ENDPOINT,
                    session: {
                        maxAge
                    }
                })
            ]
        })
    )
}


export default Auth
```

**Comment by mtt87** [comment]
Out of curiosity how do you refresh the token? Doesn’t seem like you handle it in the code posted here

**Comment by digitalhank** [comment]
yeah i removed it for brevity. my refreshAccessToken function is quite similar to the docs

**Comment by mtt87** [comment]
OT have you ever experienced issues with refresh token rotation? We are working on fixing some stuff but would be curious to hear if you ever noticed any issue with your cognito setup
https://github.com/nextauthjs/next-auth/issues/2071

**Comment by digitalhank** [comment]
Not yet. My implementation is fairly new so it hasn't been thoroughly tested yet. I can report back if anything unexpected happens.

**Comment by indianabenny** [comment]
Hi @digitalhank. I'm running into the same issue. Thank you for posting your code example above. I'm new to Next.js. Would you mind posting a quick example of how you are calling Auth in your code? Are you even using the [...nextauth].js with a <Provider> setup in _app.js or are you handling that in some other manner?

**Comment by digitalhank** [comment]
@indianabenny the only significant difference between how it's invoked in the docs and my implementation is that I am wrapping the NextAuth function with the `Auth` function as you can see above.

**Comment by dennervidal** [comment]
This workaround is pretty neat actually. I've been working with keycloak auth services, and had to apply this in my app to work properly. Only thing I say to you who had the same problem using JWT is to pay attention on session time and cookie expiration (Keycloak is not a default provider, so the management is totally on me). In order to auto refresh the tokens, set the provider options.

I think this should be considered in future versions of next-auth, since not every JWT is short enough to fit in a cookie.

**Comment by indianabenny** [comment]
See another example here for how this can be solved:
https://stackoverflow.com/a/68181560/16337541

**Comment by stale** [comment]
Hi there! It looks like this issue hasn't had any activity for a while. It will be closed if no further activity occurs. If you think your issue is still relevant, feel free to comment on it to keep it open. (Read more at #912) Thanks!

**Comment by thetaPC** [comment]
This is still an issue.

**Comment by thorep** [comment]
Im having issues storing refresh token ...if i to a substr and take just part of it , i dont get errors...

**Reply by balazsorban44** [reply]
This is a browser limitation. A single cookie can only contain 4096 bytes, if you try to set a value that exceeds it, the cookie simply won't be created. To mitigate this, we should introduce cookie chunking for the session cookie, which is not currently in the works. You could try using a database adapter, but understandably that would complicate your setup. Another way is to chunk the cookie yourself, but I only recommend that if you know what you are doing. The `jwt` callback is the right place for that. Remember if you use `refresh_token`, you should probably encrypt the JWT as well **in addition** to the default signing. Encryption might introduce a slight sizing overhead on its own.

**Comment by balazsorban44** [comment]
Update, cookie chunking is on the way! https://github.com/nextauthjs/next-auth/pull/3101 🎉

**Reply by digitalhank** [reply]
lovely to hear!
